text/0000-shared-layers.md

```diff
@@ -0,0 +1,117 @@
+
+- What is the expected outcome?
+
+The spec and lifecycle is modified to support the above use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC introduces a new top level folder structure called `<shared-layers>` which shall live on the same level as the `<layers>` directory.
+
+The structure would look like - 
+
+```
+<shared-layers> (configurable with `CNB_SHARED_LAYERS_DIR`)
+└── <escaped-buildpack-id> `CNB_BP_SHARED_LAYERS_DIR`
+    ├── <shared-layer>.toml
+    └── <shared-layer>
+        ├── <usable-directory>
+        └── env
+```
+
+Each buildpack will be passed an environment variable `CNB_BP_SHARED_LAYERS_DIR` with the location to the shared layers directory for a buildpack.
+
+The `<shared-layer>.toml` would look like - 
+
+```toml
+[types]
+  cache = false
+
+[metadata]
+# buildpack-specific data
+```
+
+The `env` directory would be used to set build time only variables for subsequent buildpacks. It will follow an API similar to [Buildpack ENV API](https://github.com/buildpacks/spec/blob/main/buildpack.md#provided-by-the-buildpacks)
+
+
+# How it Works
+[how-it-works]: #how-it-works
+
+All layers created in the `shared-layers` directory are build layers by default. The `cache` layer flag decides if the layer will be cached for the next run or not. This flag will follow the same semantics as [specified here.](https://github.com/buildpacks/spec/blob/main/buildpack.md#cached-layers). The `<shared-layer>.toml` for each layer will be restored with the same semantics as [specified here.](https://github.com/buildpacks/spec/blob/main/buildpack.md#cached-layers) with the `types` table being omitted during subsequent restores to avoid stale layers.
+
+A `shared-layer` is associated with a specific buildpack that created it and it is expected to be wiped if the associated buildpack is not involved during the build process. The `env` directory is co-located with the `shared-layer` so that the associated environment variables that may point to directories in the `shared-layer` follow the same lifecycle as the `shared-layer` itself.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+
+- Added maintanence complexity.
+- Additional complexity for buildpack authors to account for.
+
+# Alternatives
+[alternatives]: #alternatives
+
+Not implement shared layers.
+
+# Prior Art
+[prior-art]: #prior-art
+
+Related RFCs - 
+
+- https://github.com/buildpacks/rfcs/pull/155
+- https://github.com/buildpacks/rfcs/pull/145
+
+The plan would be that this `<shared-layers>` directory would be relocated alongside the `<layers>` if the proposal at https://github.com/buildpacks/rfcs/pull/145 is accepted.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- How would we accomodate for older platform APIs that may not have a mounted volume for `shared-layers`?
```

Comment:
Does it matter if it is a volume or not? As long as the `shared-layers` folder exists and is writeable, things should be fine I think.

Reply:
It will be impossible for the exporter to cache shared layers if they are not on a volume that is passed between phases. I think we have a couple options here:

1. Accept that shared layers won't be cached on older platforms and thus newer buildpacks that use this feature might be less performant on older platforms.
1. Have the builder create a shared layers dir under `layers`. It'll be ugly but given the requirement that a buildpack ID is in the form `<namespace>/<name>`, `shared` or `shared-layers` is not a valid buildpack ID and therefore there is no danger of a conflict.
1.  Have the builder copy these layers into the regular buildpack layers dir after the build is complete. This will cause some performance degradation on older platforms, but it probably isn't as dramatic as option 1.
1. Make this an optional feature of the buildpack API. e.g the platform MAY give you a shared layers dir, if not assumed `layers` is fine. This probably foists too much of the complexity onto buildpack authors.
1. Assume that by the time this is implemented the build phase will have taken over responsibility for archiving layers and putting them somewhere the exporter can read them https://github.com/buildpacks/rfcs/pull/155. But that change is a platform API change so it doesn't really solve the problem? Unless we ignore some things that are technically specified in the inter-phase API - if competing implementations of individuals phases is a non-goal, we potentially specified more than we should have...

Reply:
The volume is only required for untrusted builder paths right? `creator` flows wouldn't have to use a volume. Subsequent builds would restore into the `shared-layers` dir from the cache volume/image right?

Reply:
We should probably answer this in this RFC before it goes through.

Reply:
I went with 2. Although buildpacks published to the registry must be of the form `<namespace>/<name>` the spec doesn't prohibit buildpacks which don't have a namespace. It does however only allow alpha numeric + `./-` characters. So went with `@shared` instead. https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpacktoml-toml